README.md
---
layout: page
title: README
permalink: /README/
---
# ELMO-8

## A Simple Game Library Inspired by the Excellent PICO-8

The goal of the library is to offer a small, simple game development environment which aspires to be as charming as [PICO-8](http://www.lexaloffle.com/pico-8.php).

There are deliberate limitations to keep things simpler and to encourage the 8-bit aesthetic.

Aspirational specs:

- Display: 128x128 16 colours
- Sprites: 128 8x8 sprites
- Map: 128x32 cels
- Controls: 2 6-button joysticks

## Goals

- Be a nice little playground
- Be a simple teaching tool
- Be a nice way to write small games

## Using

Install from http://package.elm-lang.org/packages/micktwomey/elmo-8/latest

1. `elm package install micktwomey/elmo-8`
2. Edit Hello.elm
3. Download [examples/birdwatching.png](examples/birdwatching.png)
3. `elm reactor`
4. Go to http://localhost:8000/Hello.elm

Hello.elm:
```elm
import Elmo8.Console as Console
import Elmo8.Pico8 as Pico8

type alias Model = {}

draw : Model ->  List Console.Command
draw model =
    [ Console.putPixel 10 10 Pico8.peach
    , Console.print "Hello World" 10 50 Pico8.orange
    , Console.sprite 0 60 90
    ]

update : Model -> Model
update model = model

main : Program Never
main =
    Console.boot
        { draw = draw
        , init = {}
        , update = update
        , spritesUri = "birdwatching.png"
        }
```

The result should look like this:


![Basic Example](https://raw.githubusercontent.com/micktwomey/elmo-8/master/example.png)

## Examples

To play with the examples in this repo:

1. `git clone`
2. `elm package install` (not required but useful to check dependencies)
2. `elm reactor`
3. Look in http://localhost:8000/examples/

## Posts

- http://www.twoistoomany.com/blog/2016/10/19/working-on-elmo-8
- http://www.twoistoomany.com/blog/2016/10/20/elmo-8-now-with-fonts


CHANGELOG
2.0.0
- Upgrade to Elm 0.18, no changes to API.

1.0.2
- Slightly improve examples.

1.0.1
- Attempting to fix compatibility with Elm 0.17 on package index.

1.0.0
- Initial release for real (fixing numbering to match Elm's versioning system).

0.1.0
- Initial release.
- Implemented: `print`, `sprite` and `putPixel`.


tools/convert_to_data_uri.py
import argparse
import base64
import os
import sys

TEMPLATE="""module {module} exposing (..)

-- Generated from {filename}
{function}: String
{function} =
    "data:{mime_type};base64,{data}"

"""

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("module_name", help="Elm module name, e.g. Elmo8.Textures.Pico8")
    parser.add_argument("function_name", help="Elm functio name, e.g. myFontData")
    parser.add_argument("mime_type", help="mime type, e.g. image/png")
    parser.add_argument("input", help="File to read, e.g. foo.png", type=argparse.FileType('rb'))
    parser.add_argument("output", help="File to write, e.g. src/Elmo8/Textures/Pico8.elm", type=argparse.FileType('w'))
    args = parser.parse_args()

    data = base64.b64encode(args.input.read()).decode("ASCII")

    code = TEMPLATE.format(
        module=args.module_name,
        filename=os.path.basename(args.input.name),
        function=args.function_name,
        mime_type=args.mime_type,
        data=data,
    )
    args.output.write(code)

if __name__ == "__main__":
    main()


font/pico-8_regular_8.lua
return {
	file="pico-8_regular_8.PNG",
	height=10,
	description={
		family="PICO-8",
		style="Regular",
		size=8
	},
	metrics={
		ascender=10,
		descender=0,
		height=10
	},
	texture={
		file="pico-8_regular_8.PNG",
		width=128,
		height=128
	},
	chars={
		{char=" ",width=8,x=1,y=11,w=0,h=0,ox=0,oy=0},
		{char="!",width=8,x=2,y=1,w=2,h=10,ox=2,oy=10},
		{char='"',width=8,x=5,y=1,w=6,h=4,ox=0,oy=10},
		{char="#",width=8,x=12,y=1,w=6,h=10,ox=0,oy=10},
		{char="$",width=8,x=19,y=1,w=6,h=10,ox=0,oy=10},
		{char="%",width=8,x=26,y=1,w=6,h=10,ox=0,oy=10},
		{char="&",width=8,x=33,y=1,w=6,h=10,ox=0,oy=10},
		{char="'",width=8,x=40,y=1,w=4,h=4,ox=0,oy=10},
		{char="(",width=8,x=45,y=1,w=4,h=10,ox=0,oy=10},
		{char=")",width=8,x=50,y=1,w=4,h=10,ox=2,oy=10},
		{char="*",width=8,x=55,y=1,w=6,h=10,ox=0,oy=10},
		{char="+",width=8,x=62,y=3,w=6,h=6,ox=0,oy=8},
		{char=",",width=8,x=69,y=7,w=4,h=4,ox=0,oy=4},
		{char="-",width=8,x=74,y=5,w=6,h=2,ox=0,oy=6},
		{char=".",width=8,x=81,y=9,w=2,h=2,ox=2,oy=2},
		{char="/",width=8,x=84,y=1,w=6,h=10,ox=0,oy=10},
		{char="0",width=8,x=91,y=1,w=6,h=10,ox=0,oy=10},
		{char="1",width=8,x=98,y=1,w=6,h=10,ox=0,oy=10},
		{char="2",width=8,x=105,y=1,w=6,h=10,ox=0,oy=10},
		{char="3",width=8,x=112,y=1,w=6,h=10,ox=0,oy=10},
		{char="4",width=8,x=119,y=1,w=6,h=10,ox=0,oy=10},
		{char="5",width=8,x=1,y=12,w=6,h=10,ox=0,oy=10},
		{char="6",width=8,x=8,y=12,w=6,h=10,ox=0,oy=10},
		{char="7",width=8,x=15,y=12,w=6,h=10,ox=0,oy=10},
		{char="8",width=8,x=22,y=12,w=6,h=10,ox=0,oy=10},
		{char="9",width=8,x=29,y=12,w=6,h=10,ox=0,oy=10},
		{char=":",width=8,x=36,y=14,w=2,h=6,ox=2,oy=8},
		{char=";",width=8,x=39,y=14,w=4,h=8,ox=0,oy=8},
		{char="<",width=8,x=44,y=12,w=6,h=10,ox=0,oy=10},
		{char="=",width=8,x=51,y=14,w=6,h=6,ox=0,oy=8},
		{char=">",width=8,x=58,y=12,w=6,h=10,ox=0,oy=10},
		{char="?",width=8,x=65,y=12,w=6,h=10,ox=0,oy=10},
		{char="@",width=8,x=72,y=12,w=6,h=10,ox=0,oy=10},
		{char="A",width=8,x=79,y=12,w=6,h=10,ox=0,oy=10},
		{char="B",width=8,x=86,y=12,w=6,h=10,ox=0,oy=10},
		{char="C",width=8,x=93,y=12,w=6,h=10,ox=0,oy=10},
		{char="D",width=8,x=100,y=12,w=6,h=10,ox=0,oy=10},
		{char="E",width=8,x=107,y=12,w=6,h=10,ox=0,oy=10},
		{char="F",width=8,x=114,y=12,w=6,h=10,ox=0,oy=10},
		{char="G",width=8,x=1,y=23,w=6,h=10,ox=0,oy=10},
		{char="H",width=8,x=8,y=23,w=6,h=10,ox=0,oy=10},
		{char="I",width=8,x=15,y=23,w=6,h=10,ox=0,oy=10},
		{char="J",width=8,x=22,y=23,w=6,h=10,ox=0,oy=10},
		{char="K",width=8,x=29,y=23,w=6,h=10,ox=0,oy=10},
		{char="L",width=8,x=36,y=23,w=6,h=10,ox=0,oy=10},
		{char="M",width=8,x=43,y=23,w=6,h=10,ox=0,oy=10},
		{char="N",width=8,x=50,y=23,w=6,h=10,ox=0,oy=10},
		{char="O",width=8,x=57,y=23,w=6,h=10,ox=0,oy=10},
		{char="P",width=8,x=64,y=23,w=6,h=10,ox=0,oy=10},
		{char="Q",width=8,x=71,y=23,w=6,h=10,ox=0,oy=10},
		{char="R",width=8,x=78,y=23,w=6,h=10,ox=0,oy=10},
		{char="S",width=8,x=85,y=23,w=6,h=10,ox=0,oy=10},
		{char="T",width=8,x=92,y=23,w=6,h=10,ox=0,oy=10},
		{char="U",width=8,x=99,y=23,w=6,h=10,ox=0,oy=10},
		{char="V",width=8,x=106,y=23,w=6,h=10,ox=0,oy=10},
		{char="W",width=8,x=113,y=23,w=6,h=10,ox=0,oy=10},
		{char="X",width=8,x=120,y=23,w=6,h=10,ox=0,oy=10},
		{char="Y",width=8,x=1,y=34,w=6,h=10,ox=0,oy=10},
		{char="Z",width=8,x=8,y=34,w=6,h=10,ox=0,oy=10},
		{char="[",width=8,x=15,y=34,w=4,h=10,ox=0,oy=10},
		{char="\\",width=8,x=20,y=34,w=6,h=10,ox=0,oy=10},
		{char="]",width=8,x=27,y=34,w=4,h=10,ox=2,oy=10},
		{char="^",width=8,x=32,y=34,w=6,h=4,ox=0,oy=10},
		{char="_",width=8,x=39,y=42,w=6,h=2,ox=0,oy=2},
		{char="`",width=8,x=46,y=34,w=4,h=4,ox=2,oy=10},
		{char="a",width=8,x=51,y=36,w=6,h=8,ox=0,oy=8},
		{char="b",width=8,x=58,y=36,w=6,h=8,ox=0,oy=8},
		{char="c",width=8,x=65,y=36,w=6,h=8,ox=0,oy=8},
		{char="d",width=8,x=72,y=36,w=6,h=8,ox=0,oy=8},
		{char="e",width=8,x=79,y=36,w=6,h=8,ox=0,oy=8},
		{char="f",width=8,x=86,y=36,w=6,h=8,ox=0,oy=8},
		{char="g",width=8,x=93,y=36,w=6,h=8,ox=0,oy=8},
		{char="h",width=8,x=100,y=36,w=6,h=8,ox=0,oy=8},
		{char="i",width=8,x=107,y=36,w=6,h=8,ox=0,oy=8},
		{char="j",width=8,x=114,y=36,w=6,h=8,ox=0,oy=8},
		{char="k",width=8,x=1,y=47,w=6,h=8,ox=0,oy=8},
		{char="l",width=8,x=8,y=47,w=6,h=8,ox=0,oy=8},
		{char="m",width=8,x=15,y=47,w=6,h=8,ox=0,oy=8},
		{char="n",width=8,x=22,y=47,w=6,h=8,ox=0,oy=8},
		{char="o",width=8,x=29,y=47,w=6,h=8,ox=0,oy=8},
		{char="p",width=8,x=36,y=47,w=6,h=8,ox=0,oy=8},
		{char="q",width=8,x=43,y=47,w=6,h=8,ox=0,oy=8},
		{char="r",width=8,x=50,y=47,w=6,h=8,ox=0,oy=8},
		{char="s",width=8,x=57,y=47,w=6,h=8,ox=0,oy=8},
		{char="t",width=8,x=64,y=47,w=6,h=8,ox=0,oy=8},
		{char="u",width=8,x=71,y=47,w=6,h=8,ox=0,oy=8},
		{char="v",width=8,x=78,y=47,w=6,h=8,ox=0,oy=8},
		{char="w",width=8,x=85,y=47,w=6,h=8,ox=0,oy=8},
		{char="x",width=8,x=92,y=47,w=6,h=8,ox=0,oy=8},
		{char="y",width=8,x=99,y=47,w=6,h=8,ox=0,oy=8},
		{char="z",width=8,x=106,y=47,w=6,h=8,ox=0,oy=8},
		{char="{",width=8,x=113,y=45,w=6,h=10,ox=0,oy=10},
		{char="|",width=8,x=120,y=45,w=2,h=10,ox=2,oy=10},
		{char="}",width=8,x=1,y=56,w=6,h=10,ox=0,oy=10},
		{char="~",width=8,x=8,y=58,w=6,h=6,ox=0,oy=8},
		{char="À",width=16,x=15,y=56,w=14,h=10,ox=0,oy=10},
		{char="Á",width=16,x=30,y=56,w=14,h=10,ox=0,oy=10},
		{char="Â",width=16,x=45,y=56,w=14,h=10,ox=0,oy=10},
		{char="Ã",width=16,x=60,y=56,w=14,h=10,ox=0,oy=10},
		{char="Ä",width=16,x=75,y=56,w=14,h=10,ox=0,oy=10},
		{char="Å",width=16,x=90,y=56,w=10,h=10,ox=2,oy=10},
		{char="Æ",width=16,x=101,y=56,w=10,h=10,ox=2,oy=10},
		{char="Ç",width=16,x=112,y=56,w=10,h=10,ox=2,oy=10},
		{char="È",width=16,x=1,y=67,w=14,h=10,ox=0,oy=10},
		{char="É",width=16,x=16,y=67,w=10,h=10,ox=2,oy=10},
		{char="Ê",width=16,x=27,y=67,w=14,h=10,ox=0,oy=10},
		{char="Ë",width=16,x=42,y=67,w=14,h=10,ox=0,oy=10},
		{char="Ì",width=16,x=57,y=67,w=14,h=10,ox=0,oy=10},
		{char="Í",width=16,x=72,y=67,w=10,h=10,ox=2,oy=10},
		{char="Î",width=16,x=83,y=67,w=14,h=10,ox=0,oy=10},
		{char="Ï",width=16,x=98,y=67,w=10,h=10,ox=2,oy=10},
		{char="Ð",width=16,x=109,y=71,w=14,h=2,ox=0,oy=6},
		{char="Ñ",width=16,x=1,y=78,w=14,h=10,ox=0,oy=10},
		{char="Ò",width=16,x=16,y=78,w=14,h=10,ox=0,oy=10},
		{char="Ó",width=16,x=31,y=78,w=10,h=10,ox=2,oy=10},
		{char="Ô",width=16,x=42,y=78,w=14,h=10,ox=0,oy=10},
		{char="Õ",width=16,x=57,y=80,w=14,h=6,ox=0,oy=8},
		{char="Ö",width=16,x=72,y=80,w=14,h=6,ox=0,oy=8},
		{char="×",width=16,x=87,y=78,w=14,h=10,ox=0,oy=10},
		{char="Ø",width=16,x=102,y=78,w=14,h=10,ox=0,oy=10},
		{char="Ù",width=16,x=1,y=89,w=14,h=10,ox=0,oy=10},
	},
}
